dev/repos/rules.mk

```
@@ -0,0 +1,295 @@
+ONLY_${REPO_MODE}_DESCRIBE_TARGETS += describe-${REPO_NAME}
+.PHONY: describe-${REPO_NAME}
+describe-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@git -C ${REPO_DIR} log --pretty=tformat:'${REPO_DIR}: %H' -n 1
+	@git -C ${REPO_DIR} diff HEAD --quiet || echo "${REPO_DIR} is dirty"
+else
+	@echo "No repository in ${REPO_DIR}, cannot describe."
+endif
+
+FETCH_TARGETS += fetch-${REPO_NAME}
+${REPO_GROUP}_FETCH_TARGETS += fetch-${REPO_NAME}
+ONLY_${REPO_VIS}_FETCH_TARGETS += fetch-${REPO_NAME}
+ONLY_${REPO_MODE}_FETCH_TARGETS += fetch-${REPO_NAME}
+.PHONY: fetch-${REPO_NAME}
+fetch-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Fetching in ${REPO_DIR}."
+	$(Q)git -C ${REPO_DIR} fetch --all --quiet
+else
+	@echo "No repository in ${REPO_DIR}, cannot fetch."
+endif
+
+PULL_TARGETS += pull-${REPO_NAME}
+${REPO_GROUP}_PULL_TARGETS += pull-${REPO_NAME}
+ONLY_${REPO_VIS}_PULL_TARGETS += pull-${REPO_NAME}
+ONLY_${REPO_MODE}_PULL_TARGETS += pull-${REPO_NAME}
+.PHONY: pull-${REPO_NAME}
+pull-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Pulling in ${REPO_DIR}."
+	$(Q)git -C ${REPO_DIR} pull --rebase
+else
+	@echo "No repository in ${REPO_DIR}, cannot pull."
+endif
+
+PUSH_TARGETS += push-${REPO_NAME}
+${REPO_GROUP}_PUSH_TARGETS += push-${REPO_NAME}
+ONLY_${REPO_VIS}_PUSH_TARGETS += push-${REPO_NAME}
+ONLY_${REPO_MODE}_PUSH_TARGETS += push-${REPO_NAME}
+.PHONY: push-${REPO_NAME}
+push-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Pushing in ${REPO_DIR}."
+	$(Q)git -C ${REPO_DIR} push
+else
+	@echo "No repository in ${REPO_DIR}, cannot push."
+endif
+
+PEEK_TARGETS += peek-${REPO_NAME}
+${REPO_GROUP}_PEEK_TARGETS += peek-${REPO_NAME}
+ONLY_${REPO_VIS}_PEEK_TARGETS += peek-${REPO_NAME}
+ONLY_${REPO_MODE}_PEEK_TARGETS += peek-${REPO_NAME}
+.PHONY: peek-${REPO_NAME}
+peek-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Peeking into ${REPO_DIR}:"
+	@git -C ${REPO_DIR} status --short --branch --untracked-files=normal
+else
+	@echo "No repository in ${REPO_DIR}, cannot peek."
+endif
+
+GITCLEAN_TARGETS += gitclean-${REPO_NAME}
+${REPO_GROUP}_GITCLEAN_TARGETS += gitclean-${REPO_NAME}
+ONLY_${REPO_VIS}_GITCLEAN_TARGETS += gitclean-${REPO_NAME}
+ONLY_${REPO_MODE}_GITCLEAN_TARGETS += gitclean-${REPO_NAME}
+.PHONY: gitclean-${REPO_NAME}
+gitclean-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Cleaning ${REPO_DIR}:"
+	$(Q)git -C ${REPO_DIR} clean -xfd
+else
+	@echo "No repository in ${REPO_DIR}, cannot clean."
+endif
+
+CHECKOUT_MAIN_TARGETS += checkout-main-${REPO_NAME}
+${REPO_GROUP}_CHECKOUT_MAIN_TARGETS += checkout-main-${REPO_NAME}
+ONLY_${REPO_VIS}_CHECKOUT_MAIN_TARGETS += checkout-main-${REPO_NAME}
+ONLY_${REPO_MODE}_CHECKOUT_MAIN_TARGETS += checkout-main-${REPO_NAME}
+.PHONY: checkout-main-${REPO_NAME}
+checkout-main-${REPO_NAME}:
+ifeq ($(wildcard ${REPO_DIR}),${REPO_DIR})
+	@echo "Resetting ${REPO_DIR} to ${REPO_DEFAULT}:"
```

This "resetting" isn't a `git reset`, so for clarity I'd prefer

```suggestion
	@echo "Switching ${REPO_DIR} to default branch ${REPO_DEFAULT}:"
```
or
```suggestion
	@echo "Checking out default branch ${REPO_DEFAULT} in ${REPO_DIR}:"
```

---

Opting for a variation of the second option.